iib/workers/tasks/build.py

```diff
@@ -337,6 +342,146 @@ def _get_resolved_image(pull_spec):
+    :return: tuple containing port number of the running service and the running Popen object.
+    :rtype: (int, Popen)
+    :raises IIBError: if all tried ports are in use, or the command failed for another reason.
+    """
+    conf = get_worker_config()
+    port_start = conf['iib_grpc_start_port']
+    port_end = port_start + conf['iib_grpc_max_port_tries']
+
+    for port in range(port_start, port_end):
+        try:
+            return (
+                port,
+                _serve_index_registry_at_port(
+                    db_path, port, conf['iib_grpc_max_tries'], conf['iib_grpc_init_wait_time']
+                ),
+            )
+        except AddressAlreadyInUse:
+            log.info('Port %d is in use, trying another.', port)
+
+    err_msg = f'No free port has been found after {conf.get("iib_grpc_max_port_tries")} attempts.'
+    log.error(err_msg)
+    raise IIBError(err_msg)
+
+
+def _serve_index_registry_at_port(db_path, port, max_tries, wait_time):
+    """
+    Start an image registry service at a specified port.
+
+    :param str db_path: path to index database containing the registry data.
+    :param str int port: port to start the service on.
+    :param max_tries: how many times to try to start the service before giving up.
+    :param wait_time: time to wait before checking if the service is initialized.
+    :return: object of the running Popen process.
+    :rtype: Popen
+    :raises IIBError: if the process has failed to initialize too many times, or an unexpected
+        error occured.
+    :raises AddressAlreadyInUse: if the specified port is already being used by another service.
+    """
+    cmd = ['opm', 'registry', 'serve', '-p', str(port), '-d', db_path]
+    for attempt in range(max_tries):
+        rpc_proc = subprocess.Popen(
+            cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True
+        )
+        start_time = time.time()
+        while time.time() - start_time < wait_time:
+            time.sleep(1)
+            ret = rpc_proc.poll()
+            # process has terminated
+            if ret is not None:
+                stderr = rpc_proc.stderr.read()
+                if 'address already in use' in stderr:
+                    raise AddressAlreadyInUse(f'Port {port} is already used by a different service')
+                raise IIBError(f'Command "{" ".join(cmd)}" has failed with error "{stderr}"')
+
+            stdout = get_running_subprocess_output(rpc_proc.stdout)
+            if 'serving registry' in stdout:
+                log.debug('Started the command "%s"', ' '.join(cmd))
+                log.info('Index registry service has been initialized.')
+                return rpc_proc
+
+        rpc_proc.kill()
+
+    raise IIBError(f'Index registry has not been initialized after {max_tries} tries')
+
+
+@retry(attempts=2, wait_on=IIBError, logger=log)
+def _get_present_bundles(from_index, base_dir):
```

Comment:
Instead of having a retry here, I would rather have it on `_serve_index_registry_at_port` and `_serve_index_registry`

Reply:
I have removed this retry and added a retry to _serve_index_registry_at_port. However, I don't see any operations in _serve_index_registry that could benefit from a retry.